software/contrib/cvecorder.py

```diff
@@ -0,0 +1,348 @@
+            elif ticks_diff(ticks_ms(), b2.last_pressed()) >  500:
+                self.clearCvrs(self.ActiveBank)
+                self.bankToSave = self.ActiveBank
+                self.saveState()
+                # reverse the ActiveCvr increment caused by the initial button press
+                if self.ActiveCvr > 0:
+                    self.ActiveCvr -= 1
+                else:
+                    self.ActiveCvr = self.numCVR
+
+        # B2 short press
+        @b2.handler
+        def b2Pressed():
+            # Change the active recorder channel
+            if self.ActiveCvr < self.numCVR:
+                self.ActiveCvr += 1
+            else:
+                self.ActiveCvr = 0
+
+    def handleClock(self):
+
+        # Sample input to 2 decimal places
+        self.CvIn = round(20 * ain.percent(), 2)
+
+        # Start recording if pending and on first step
+        if self.step == 0 and self.CvRecording[self.ActiveCvr] == 'pending':
+            self.CvRecording[self.ActiveCvr] = 'true'
+
+        for i in range(self.numCVR+1):
+            # If recording, write the sampled value to the CVR list and play the voltage
+            if self.CvRecording[i] == 'true':
+                self.CVR[self.ActiveBank][self.ActiveCvr][self.step] = self.CvIn
+                cvs[self.ActiveCvr].voltage(self.CvIn)
+            else:
+                cvs[i].voltage(self.CVR[self.ActiveBank][i][self.step])
+
+        # Reset step number at stepLength -1 as pattern arrays are zero-based
+        if self.step < self.stepLength - 1:
+            self.step += 1
+        else:
+            # Reset step to zero , stop recording and save recording to local storage
+            self.step = 0
+            if self.CvRecording[self.ActiveCvr] == 'true':
+                self.CvRecording[self.ActiveCvr] = 'false'
+                self.bankToSave = self.ActiveBank
+                self.saveState()
+
+    def clearCvrs(self, bank):
+        for b in range(self.numCVRBanks+1):
+            # skip bank unless 'all' is passed
+            if b != bank and bank != 'all':
+                continue
+            if self.debugTest:
+                print('Clearing bank: ' + str(b))
+            # Set all CV values to zero
+            for i in range(self.numCVR+1):
+                for n in range (0, self.stepLength):
+                    self.CVR[b][i][n] = 0
+            # Save the cleared bank to local storage
+            self.bankToSave = b
+            self.saveState()
+
+    def initCvrs(self):
+        for b in range(self.numCVRBanks+1):
+            self.CVR.append([])
+            for i in range(self.numCVR+1):
+                self.CVR[b].append([])
+                self.CvRecording.append('false')
+                for n in range (0, self.stepLength):
+                    self.CVR[b][i].append(0)
+
+    def saveState(self):
+        # generate output filename
+        outputFile = f"saved_state_{self.__class__.__qualname__}_{self.bankToSave}.txt"
```

Comment:
Why is `bank_to_save` a member variable instead of a method parameter?

Reply:
Apologies, I don't understand the question. Would you mind elaborating and also explaining the difference, thanks.

Reply:
Sorry, typo, it's `bankToSave` not `bank_to_save`

`bankToSave` is a member variable of the class `CVecoder`, that is, it is a field on the class. But I don't think it needs to be. Every time you set `self.bankToSave` you immediately call `self.saveState()`, and you never call `saveState()` without setting `bankToSave` first. They are intimately linked. I don't see any reason why `bankToSave` should be a field on the class, instead it should be a parameter on the `saveState()` method. As in:

```Python
def saveState(self, bankToSave):
    # generate output filename
    outputFile = f"saved_state_{self.__class__.__qualname__}_{bankToSave}.txt"
    
    ...
```

And the call site, at line 181, for example:
```Python
# Save the cleared bank to local storage
self.saveState(b)
```

Does that make sense?

Reply:
oh, wow, got it. Let's add some documentation to describe this. Maybe at the beginning of the block of vars that are initialized for this purpose in the `__init__`?

Reply:
Done! thanks for the suggestion.

Reply:
New doc looks good.